### PR TITLE
WFP 2728 backend unallocated cases screen

### PR DESCRIPTION
**Pull request by GaReid**

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsallocations/client/WorkforceAllocationsToDeliusApiClient.kt [hunk 1]
```diff
@@ -38,6 +40,28 @@ class WorkforceAllocationsToDeliusApiClient(private val webClient: WebClient) {
     private val policy: PolicyFactory = Sanitizers.FORMATTING.and(Sanitizers.LINKS)
   }
 
+  suspend fun getApopUsers(): List<DeliusApopUser> {
```
Comment (wood-n) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsallocations/client/WorkforceAllocationsToDeliusApiClient.kt:43]: Can we add a comment here to specify what the endpoint is supposed to get?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsallocations/client/WorkforceAllocationsToDeliusApiClient.kt [hunk 2]
```diff
@@ -38,6 +40,28 @@ class WorkforceAllocationsToDeliusApiClient(private val webClient: WebClient) {
...
+      .awaitBody()
+  }
+  suspend fun getAllUserAccessByCrn(crn: String): DeliusLimitedAccessDetails {
+    return webClient
+      .get()
+      .uri("person/$crn/limited-access/all")
+      .retrieve()
+      .awaitBody()
+  }
+  suspend fun geUserAccessByCrnUsers(crn: String, staffCodes: List<String>): DeliusLimitedAccessDetails {
```
Comment (wood-n) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsallocations/client/WorkforceAllocationsToDeliusApiClient.kt:57]: getStaffAccessRestrictionsForCrn()

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsallocations/client/WorkforceAllocationsToDeliusApiClient.kt [hunk 3]
```diff
@@ -38,6 +40,28 @@ class WorkforceAllocationsToDeliusApiClient(private val webClient: WebClient) {
...
   }
 
+  suspend fun getApopUsers(): List<DeliusApopUser> {
+    return webClient
+      .get()
+      .uri("/users")
+      .retrieve()
+      .awaitBody()
+  }
+  suspend fun getAllUserAccessByCrn(crn: String): DeliusLimitedAccessDetails {
```
Comment (wood-n) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsallocations/client/WorkforceAllocationsToDeliusApiClient.kt:50]: getUserAccessRestrictionsByCrn()

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsallocations/client/dto/DeliusLimitedAccessDetails.kt [hunk 4]
```diff
@@ -0,0 +1,9 @@
+package uk.gov.justice.digital.hmpps.hmppsallocations.client.dto
+
+data class DeliusLimitedAccessDetails constructor(
```
Comment (wood-n) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsallocations/client/dto/DeliusLimitedAccessDetails.kt:3]: DeliusAccessRestrictionDetails

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsallocations/service/LaoService.kt [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+
+@Service
+class LaoService(
+  @Qualifier("workforceAllocationsToDeliusApiClientUserEnhanced")
+  private val workforceAllocationsToDeliusApiClient: WorkforceAllocationsToDeliusApiClient,
+) {
+
+  suspend fun getCrnRestrictions(crn: String): DeliusCrnRestrictions {
+    val apopUsers = workforceAllocationsToDeliusApiClient.getApopUsers()
+    val staffCodes = apopUsers.map { it.staffCode }
```
Comment (wood-n) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsallocations/service/LaoService.kt:16]: apopUsersStaffCodes

Comment (GaReid) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsallocations/service/LaoService.kt]: Have done the suggested refactors and a couple of others.

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsallocations/integration/unallocatedcases/GetUnallocatedCasesByTeamTests.kt [hunk 6]
```diff
@@ -154,6 +163,11 @@ class GetUnallocatedCasesByTeamTests : IntegrationTestBase() {
         Triple("ZZZZZZZ", true, true),
```
Comment (wood-n) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsallocations/integration/unallocatedcases/GetUnallocatedCasesByTeamTests.kt:163]: Update  Get unallocated cases by team where probation-estate API is successful and does not return LAO cases to change LAO to restricted

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsallocations/integration/unallocatedcases/GetUnallocatedCasesByTeamTests.kt [hunk 7]
```diff
@@ -154,6 +163,11 @@ class GetUnallocatedCasesByTeamTests : IntegrationTestBase() {
...
       ),
```
Comment (wood-n) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsallocations/integration/unallocatedcases/GetUnallocatedCasesByTeamTests.kt:164]: workforceAllocationsToDelius.userHasAccessToAllCases Method may need updating in test   `Get unallocated cases by team where all cases are LAO cases`() {

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsallocations/integration/unallocatedcases/GetUnallocatedCasesByTeamTests.kt [hunk 8]
```diff
@@ -378,6 +406,11 @@ class GetUnallocatedCasesByTeamTests : IntegrationTestBase() {
...
     )
 
+    workforceAllocationsToDelius.setExcludedUsersByCrn(
+      listOf("J678910", "J680660", "X4565764", "J680648", "X6666222", "XXXXXXX", "ZZZZZZZ", "AB77711"),
+    )
+    workforceAllocationsToDelius.setApopUsers()
+
     webTestClient.get()
       .uri("/team/TEAM1/cases/unallocated")
       .headers { it.authToken(roles = listOf("ROLE_MANAGE_A_WORKFORCE_ALLOCATE")) }
```
Comment (wood-n) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsallocations/integration/unallocatedcases/GetUnallocatedCasesByTeamTests.kt:416]: Suggest to add a test to specify new behaviour of getting excluded cases

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsallocations/service/GetUnallocatedCaseServiceTest.kt [hunk 9]
```diff
@@ -69,7 +73,36 @@ internal class GetUnallocatedCaseServiceTest {
...
+      teamCode = "TM1",
+      tier = "C2",
+      id = 2L,
+      convictionNumber = 1,
+    )
+
+    every { mockRepo.findByTeamCode("TM1") } returns listOf(unallocatedCaseEntity)
+    coEvery { mockWorkforceAllocationsToDeliusApiClientClient.getUserAccess(listOf(crn)) } returns
+      DeliusUserAccess(
+        access = listOf(DeliusCaseAccess(crn, userRestricted = true, false)),
```
Comment (wood-n) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsallocations/service/GetUnallocatedCaseServiceTest.kt:99]: Suggesting either adding parameter names for all or none

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsallocations/service/LaoServiceTest.kt [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+import uk.gov.justice.digital.hmpps.hmppsallocations.client.dto.DeliusAccessRestrictionDetails
+import uk.gov.justice.digital.hmpps.hmppsallocations.client.dto.DeliusApopUser
+
+class LaoServiceTest {
+
+  @MockK
+  lateinit var workforceAllocationsToDeliusApiClient: WorkforceAllocationsToDeliusApiClient
+
+  @InjectMockKs
+  lateinit var cut: LaoService
```
Comment (wood-n) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsallocations/service/LaoServiceTest.kt:20]: Please rename to laoService

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsallocations/service/LaoServiceTest.kt [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+
+  @BeforeEach
+  fun setUp() {
+    MockKAnnotations.init(this, relaxUnitFun = true)
+  }
+
+  @Test
+  fun `non lao case`() = runTest {
```
Comment (wood-n) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsallocations/service/LaoServiceTest.kt:28]: please make test names more verbose and descriptive

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsallocations/service/TelemetryServiceTest.kt [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+  }
+
+  @Test
+  fun trackUnallocatedCaseAllocated() {
+    val crn = "X1234567"
+    val name = "Bob Jones"
+    val teamCode = "N54ERT"
+    val providerCode = "PC001"
+    val tier = "C2"
+    val uce = UnallocatedCaseEntity(1L, name, crn, tier, teamCode, providerCode, ZonedDateTime.now(), 1)
```
Comment (wood-n) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsallocations/service/TelemetryServiceTest.kt:33]: please change from acronym

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsallocations/service/GetUnallocatedCaseServiceTest.kt (hunk 9 above)
Comment (GaReid) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsallocations/service/GetUnallocatedCaseServiceTest.kt]: added all

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsallocations/service/LaoServiceTest.kt (hunk 10 above)
Comment (GaReid) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsallocations/service/LaoServiceTest.kt]: done

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsallocations/integration/unallocatedcases/GetUnallocatedCasesByTeamTests.kt (hunk 8 above)
Comment (GaReid) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsallocations/integration/unallocatedcases/GetUnallocatedCasesByTeamTests.kt:496]: done, test shows an excluded, an excluded that has an Apop jser  and ignores the remaining  restricted cases

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsallocations/integration/unallocatedcases/GetUnallocatedCasesByTeamTests.kt (hunk 7 above)
Comment (GaReid) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsallocations/integration/unallocatedcases/GetUnallocatedCasesByTeamTests.kt:248]: Done. Call wasn't needed in that instance

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsallocations/integration/unallocatedcases/GetUnallocatedCasesByTeamTests.kt (hunk 6 above)
Comment (GaReid) [reply src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsallocations/integration/unallocatedcases/GetUnallocatedCasesByTeamTests.kt:247]: done

**Review by wood-n — APPROVED** [review]

**Pull request merged**